beetsplug/tidal.py

```diff
@@ -0,0 +1,1050 @@
+        # and we only really need the HTTP response code anyways.
+        resp = requests.head(url)
+        try:
+            resp.raise_for_status()
+            return True
+        except requests.exceptions.HTTPError:
+            return False
+
+    def _parse_copyright(self, copyright):
+        """Attempts to extract a record label from a freeform
+        TIDAL copyright string.
+
+        :param copyright: TIDAL copyright string
+        :type copyright: str
+        :return: Record label if found, otherwise a blank string
+        :rtype: str
+        """
+        # This isn't 100% needed, but it makes calling it easier
+        if not copyright:
+            return ""
+
+        regx = re.findall(self.copyright_regex, copyright)
+        if not regx:
+            self._log.warn(
+                (
+                    "Copyright regex returned no results but "
+                    "we have a copyright, please make a bug report with `beets -vv`."
+                )
+            )
+            self._log.debug(f"Copyright: {copyright}")
+            return ""
+
+        # The last group, if multiple were found, tends to be the correct one.
+        return regx[-1]
+
+    def _track_to_trackinfo(self, track, album=None):
+        """Converts a TIDAL track to a beets TrackInfo
+
+        :param track: A track obtained from the TIDAL API
+        :type track: tidalapi.media.Track
+        :param album: Fills in optional track info, defaults to None
+        :type album: tidalapi.media.Album, optional
+        :return: A beets TrackInfo created with the provided data
+        :rtype: beets.autotag.hooks.TrackInfo
+        """
+        # Create basic trackinfo with standard fields
+        trackinfo = TrackInfo(
+            title=track.name,
+            track_id=track.id,
+            artist=track.artist.name,
+            artist_id=track.artist.id,
+            album=track.album.name,
+            length=track.duration,
+            medium=track.volume_num,
+            medium_index=track.track_num,
+            index=track.track_num,
+            data_source=self.data_source,
+            data_url=track.share_url,
+            isrc=track.isrc,
+            artists=[artist.name for artist in track.artists],
+            artists_ids=[str(artist.id) for artist in track.artists],
+            label=self._parse_copyright(track.copyright),
+        )
+
+        # Add TIDAL specific metadata
+        trackinfo.tidal_track_id = track.id
+        trackinfo.tidal_track_popularity = track.popularity  # Range: 0 to 100
+        trackinfo.tidal_artist_id = track.artist.id
+        trackinfo.tidal_album_id = track.album.id
+
+        # If we're given an album, add it's data to the track.
+        # Tidal does NOT return a lot of info on searches to save on bandwidth.
+        if album:
+            trackinfo.medium_total = album.num_tracks
+
+        return trackinfo
+
+    def _search_from_metadata(self, item, limit=10):
```

Comment:
It seems to me that this method will be slow to return results due to the number of search requests it makes. Have you tested it?